_posts/sean-has-a-crystal-ball.md

```diff
@@ -0,0 +1,110 @@
+---
+layout: post
+header_image: /img/posts/sean-has-a-crystal-ball/header.jpg
+title: "Sean has a crystal ball"
+author: Mauro Servienti
+synopsis: "Eventual consistency is everywhere. Distributed systems are not any different; in fact, they tend to make consistency issues stand out even more. Most of the time that's not a problem at all, except when it comes to the UI. What options do we have?"
+tags:
+- SOA
+- UX
+- ViewModel Composition
+---
+
+Eleven years ago, Udi Dahan wrote that ["Race Conditions Don't Exist"](http://udidahan.com/2010/08/31/race-conditions-dont-exist/); more recently, he gave a talk saying that [commands don't fail](https://www.youtube.com/watch?v=fWU8ZK0Dmxs).
+
+Along the same line, I wrote the ["Businesses don't fail, they make mistakes"](https://milestone.topics.it/2019/09/10/businesses-dont-fail-they-make-mistakes.html) article, about the fact that exceptions don't exist, and more recently in ["Tales of a reservation"](https://milestone.topics.it/2021/05/05/tales-of-a-reservation.html) I touched on the commands don't fail concept.
+
+All this is nice and hopefully useful; however, it's not enough; there's the elephant in the room, and we're ignoring it.
+
+## How do we deal with the user interface?
+
+All the topics covered in the mentioned articles and videos are about the back-end part of the system. Everything, or mostly everything, can be asynchronous and, if well designed, ignore most of the eventual consistency issues we can imagine.
+
+That's not the case when it comes to the user interface. A user in front of the screen is waiting for a response; we need to deal with the back-end's asynchronous nature not to feel lost.
+
+Let me give you two examples, diametrically opposite.
+
+First, the Amazon checkout process terminates with a screen that thanks you for your order and provides you with an order number. Then, within seconds, if not faster, an email message arrives in your inbox with the order details. At this point, I'm pretty sure that none of the order processing has happened yet. It occurred to me a couple of times that a day after having received the order placed email, the order was bounced back, for example, because the ordered item was out of stock, and there was no indication of when it would have been available again. Remember, commands don't fail. If we navigate to the order details page, even if none of the processing has happened yet, we're presented with a nice looking set of details; all those details, in my opinion, serve only a single purpose: generate confidence in customers.
```

Comment:
```suggestion
First, the Amazon checkout process terminates with a screen that thanks you for your order and provides you with an order number. Then, within seconds, if not faster, an email message arrives in your inbox with the order details. At this point, I'm pretty sure that none of the order processing has happened yet. It occurred to me a couple of times that a day after having received the "order placed" email, the order was bounced back, for example, because the ordered item was out of stock, and there was no indication of when it would have been available again. Remember, commands don't fail. If we navigate to the order details page, even if none of the processing has happened yet, we're presented with a nice looking set of details. All those details, in my opinion, serve only a single purpose: generate confidence in customers.
```